database/user.go

```diff
@@ -151,7 +153,36 @@ func (db *DB) UserBySub(ctx context.Context, sub string, create bool) (*User, er
 	users, err := db.managedUsersByField(ctx, "sub", sub)
 	if create && errors.Contains(err, ErrUserNotFound) {
 		var u *User
-		u, err = db.UserCreate(ctx, sub, TierFree)
+		_, email, err := jwt.UserDetailsFromJWT(ctx)
+		if err != nil {
+			// Log the error but don't do anything differently.
+			db.staticLogger.Debugf("We failed to extract the expected user infotmation from the JWT token. Error: %s", err.Error())
+		}
+		u, err = db.UserCreate(ctx, email, "", sub, TierFree)
+		// If we're successful or hit any error, other than a duplicate key we
+		// want to just return. Hitting a duplicate key error means we ran into
+		// a race condition and we can easily recover from that.
+		if err == nil || !strings.Contains(err.Error(), "E11000 duplicate key error collection") {
+			return u, err
+		}
+		// Recover from the race condition by fetching the existing user from
+		// the DB.
+		users, err = db.managedUsersByField(ctx, "sub", sub)
+	}
+	if err != nil {
+		return nil, err
+	}
+	return users[0], nil
+}
+
+// UserByEmail returns the user with the given username. If `create` is `true`
+// it will create the user if it doesn't exist.
+func (db *DB) UserByEmail(ctx context.Context, email string, create bool) (*User, error) {
```

Comment:
REQ: I don't see this tested.

Reply:
It seems like a good moment to overhaul the integration test suite and add tests for the uncovered methods. Tests added.